### PR TITLE
Flexer Tests

### LINKED ISSUE #981: Flexer Tests
### Summary
The current flexer implementation is somewhat lacking in tests of critical code-paths. These should be added so that we can ensure that things are working properly before we start to build on it from scratch.

### Value
We will have a more robust regression suite for if we need to modify the flexer in the future.

### Specification

- [ ] Create some tests that build the NFA from a more complex set of groups, including situations where we have multiple groups. This is useful for modelling the behaviour of a real lexer which may have multiple sets of groups.
- [ ] Add a second large-scale benchmark (like `group::test::bench_hundred_rules`) but with more complex rules (it currently only uses `Pattern::all`), and utilising more complex patterns. I'm interested in how the conversion performance scales with the complexity of patterns, or whether it's effectively linear in the number of rules.

### Acceptance Criteria & Test Cases

- We have much more comprehensive tests for the flexer, and a better idea of its performance characteristics.

### PR DESCRIPTION
**Pull request by Kesanov**
### Pull Request Description

Benchmarks NFA conversion on more complex input. Closes #981 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by iamrecursion** [comment]
Can you please resolve the conflicts so I can review this properly?

File: lib/rust/flexer/src/automata/dfa.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ use crate::data::matrix::Matrix;
 ///
 /// ```text
 ///  ┌───┐  'D'  ┌───┐  'F'  ┌───┐  'A'  ┌───┐
-///  │ 0 │──────▶│ 1 │──────▶│ 2 │──────▶│ 3 │
+///  │ 0 │ ----> │ 1 │ ----> │ 2 │ ----> │ 3 │
```
Comment (wdanilo) [inline lib/rust/flexer/src/automata/dfa.rs:19]: Why these changes? unicode arrows look so much better!

File: lib/rust/flexer/src/automata/nfa.rs [hunk 2]
```diff
@@ -38,9 +38,9 @@ type StateSetId = BTreeSet<state::Identifier>;
 /// [epsilon links](https://en.wikipedia.org/wiki/Nondeterministic_finite_automaton#NFA_with_%CE%B5-moves).
 ///
 /// ```text
-///  ┌───┐  'N'  ┌───┐   ┌───┐  'F'  ┌───┐   ┌───┐  'A'  ┌───┐
-///  │ 0 │──────▶│ 1 │──▶│ 2 │──────▶│ 3 │──▶│ 3 │──────▶│ 3 │
-///  └───┘       └───┘ ε └───┘       └───┘ ε └───┘       └───┘
+///  ┌───┐  'N'  ┌───┐    ┌───┐  'F'  ┌───┐    ┌───┐  'A'  ┌───┐
+///  │ 0 │ ----> │ 1 │ -> │ 2 │ ----> │ 3 │ -> │ 3 │ ----> │ 3 │
+///  └───┘       └───┘ ε  └───┘       └───┘ ε  └───┘       └───┘
```
Comment (wdanilo) [inline lib/rust/flexer/src/automata/nfa.rs:43]: why these changes?

**Review by wdanilo — APPROVED** [review]

File: lib/rust/flexer/src/automata/dfa.rs (hunk 1 above)
Comment (Kesanov) [reply lib/rust/flexer/src/automata/dfa.rs:19]: Yes, but they are not monospaced, so they break the alignment of boxes.

File: lib/rust/flexer/src/automata/nfa.rs (hunk 2 above)
Comment (Kesanov) [reply lib/rust/flexer/src/automata/nfa.rs:43]: Unicode arrows are not monospaced, so they break the alignment of boxes.

File: lib/rust/flexer/src/automata/dfa.rs (hunk 1 above)
Comment (iamrecursion) [reply lib/rust/flexer/src/automata/dfa.rs:19]: That is an issue with your font. They are perfectly monospaced on my end.

Comment (Kesanov) [reply lib/rust/flexer/src/automata/dfa.rs:19]: It's an issue of your font as well, because neither mine nor github monospaced fonts show them aligned at all.

**Review by iamrecursion — APPROVED** [review]
